script.module.cache/lib/cache.py

```diff
@@ -0,0 +1,311 @@
+class Store(object):
+    """
+    Generic unique string storage helper
+    Saves unique strings in a list that can
+    be retrieved, appended to, removed from or cleared entirely
+    """
+
+    def __init__(self, key, db=None):
+        # type: (str, str) -> None
+        """
+        Creates a new Store object
+        key is the identifier that strings are stored under
+        db is the path to a sqlite3 database (defaults to Cache default)
+        """
+        self.db = db
+        self.key = key
+
+    def retrieve(self):
+        # type: () -> list
+        """Gets stored strings as list"""
+        with Cache(self.db) as c:
+            data = c.get(self.key)
+            return json.loads(data["blob"]) if data else []
+
+    def _save(self, data):
+        # type: (list) -> None
+        """Saves stored data"""
+        with Cache(self.db) as c:
+            c.set(self.key, json.dumps(data, ensure_ascii=False))
+
+    def append(self, data):
+        # type: (str) -> None
+        """Add string to the store if it doesn't exist"""
+        current = self.retrieve()
+        if data not in current:
+            current.append(data)
+            self._save(current)
+
+    def remove(self, data):
+        # type: (str) -> None
+        """Remove string from the store if it exists"""
+        current = self.retrieve()
+        if data in current:
+            current.remove(data)
+            self._save(current)
+
+    def clear(self):
+        # type: () -> None
+        """Clears the store of all data"""
+        with Cache(self.db) as c:
+            c.delete(self.key)
+
+
+class GMT(tzinfo):
+    """GMT Time Zone"""
+
+    def utcoffset(self, dt):
+        return timedelta(0)
+
+    def tzname(self, dt):
+        return "GMT"
+
+    def dst(self, dt):
+        return timedelta(0)
+
+
+class Blob(object):
+    """Blob serialisation class"""
+
+    def __init__(self, data):
+        self.data = data
+
+    def __conform__(self, protocol):
+        if protocol is sqlite3.PrepareProtocol:
+            return sqlite3.Binary(cpickle.dumps(self.data, -1))
+
+    @staticmethod
+    def deserialise(data):
+        # type: (bytes) -> Any
+        return cpickle.loads(bytes(data))
+
+
+class Cache(object):
+    """
+    HTTP control directive based cache
+    https://docs.python.org/2/library/sqlite3.html
+    https://tools.ietf.org/html/rfc7234
+    """
+
+    def __init__(self, db=None):
+        # type: (str) -> None
+        """
+        Creates a new Cache object
+        db is the path to a sqlite3 database (defaults to CACHE_URI)
+        """
+        self.connection = None
+        self.db = CACHE_URI if db is None else db
+        if not xbmcvfs.exists(self.db):
+            xbmcvfs.mkdirs(path.dirname(self.db))
+        self._open(self.db)
+
+    def get(self, uri):
+        # type: (str) -> Union[None, sqlite3.Row]
+        """Retrieve a partial entry from the cache"""
+        query = ("SELECT blob, last_modified, etag, immutable, "
```

Comment:
Consider using multiline strings `"""..."""` for SQL queries.

Reply:
Done